### PR TITLE
SPARK-5087. [YARN] Merge yarn.Client and yarn.ClientBase

### PR DESCRIPTION
**Pull request by sryza**

### CODE REVIEW

**Comment by sryza** [comment]
@andrewor14 @tgravescs

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -17,34 +17,56 @@
...
- * Version of [[org.apache.spark.deploy.yarn.ClientBase]] tailored to YARN's stable API.
- */
 private[spark] class Client(
     val args: ClientArguments,
     val hadoopConf: Configuration,
     val sparkConf: SparkConf)
-  extends ClientBase with Logging {
+  extends Logging {
+
+  import Client._
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: what is this import for?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -105,25 +127,498 @@ private[spark] class Client(
...
+      throw new SparkException("Application is killed")
+    }
+    if (finalApplicationStatus == FinalApplicationStatus.UNDEFINED) {
+      throw new SparkException("The final status of application is undefined")
+    }
+  }
 }
 
-object Client {
+private[spark] object Client extends Logging {
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I'm not sure we can change this to private at this point, its already been out there are public.

Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: yes, we need to keep the object public, per discussions here #2350

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: for helper methods defined in `object Client` I believe

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 3]
```diff
@@ -17,34 +17,56 @@
...
+  import Client._
 
   def this(clientArgs: ClientArguments, spConf: SparkConf) =
     this(clientArgs, SparkHadoopUtil.get.newConfiguration(spConf), spConf)
 
   def this(clientArgs: ClientArguments) = this(clientArgs, new SparkConf())
 
-  val yarnClient = YarnClient.createYarnClient
-  val yarnConf = new YarnConfiguration(hadoopConf)
+  private val yarnClient = YarnClient.createYarnClient
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: does this call have a side effect? If it's not just a getter we need to put `()` at the end

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 4]
```diff
@@ -17,34 +17,56 @@
...
+  private val yarnConf = new YarnConfiguration(hadoopConf)
+  private val credentials = UserGroupInformation.getCurrentUser.getCredentials
+  private val amMemoryOverhead = args.amMemoryOverhead // MB
+  private val executorMemoryOverhead = args.executorMemoryOverhead // MB
+  private val distCacheMgr = new ClientDistributedCacheManager()
+  private val isLaunchingDriver = args.userClass != null
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: mind renaming this `isClusterMode` while you're at it?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 5]
```diff
@@ -105,25 +127,498 @@ private[spark] class Client(
...
-  override def setupSecurityToken(amContainer: ContainerLaunchContext): Unit = {
+  private def setupSecurityToken(amContainer: ContainerLaunchContext): Unit = {
     val dob = new DataOutputBuffer
     credentials.writeTokenStorageToStream(dob)
     amContainer.setTokens(ByteBuffer.wrap(dob.getData))
   }
 
   /** Get the application report from the ResourceManager for an application we have submitted. */
-  override def getApplicationReport(appId: ApplicationId): ApplicationReport =
+  def getApplicationReport(appId: ApplicationId): ApplicationReport =
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: can this be private?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 6]
```diff
@@ -105,25 +127,498 @@ private[spark] class Client(
...
+    // We could add checks to make sure the entire cluster has enough resources but that involves
+    // getting all the node reports and computing ourselves.
+  }
+
+  /**
+   * Copy the given file to a remote file system (e.g. HDFS) if needed.
+   * The file is only copied if the source and destination file systems are different. This is used
+   * for preparing resources for launching the ApplicationMaster container. Exposed for testing.
+   */
+  def copyFileToRemote(
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: should these just go into `object Client` too? (How many extra parameters do we need to pass them in?) If possible I'd prefer to keep the class itself smaller

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 7]
```diff
@@ -105,25 +127,498 @@ private[spark] class Client(
...
+
+    env
+  }
+
+  /**
+   * Set up a ContainerLaunchContext to launch our ApplicationMaster container.
+   * This sets up the launch environment, java options, and the command for launching the AM.
+   */
+  private def createContainerLaunchContext(newAppResponse: GetNewApplicationResponse)
+    : ContainerLaunchContext = {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: bump 1 line?

**Comment by andrewor14** [comment]
@sryza looks pretty straightforward to me. I notice there are a couple methods that aren't fully private only because they're exposed for testing. If possible, could we make these `private[yarn]` at the least? Also, have you had a chance to test this on a YARN cluster?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Yup.  We used to `import ClientBase._`

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 3 above)
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Checked in the Hadoop source and it doesn't have side effects.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 6 above)
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: It seems like moving over `copyFileToRemote` and `getQualifiedLocalPath` should be reasonable.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Do you mean add a blank line after `: ContainerLaunchContext = {`?

**Comment by sryza** [comment]
Thanks for taking a look Andrew and Tom.  I tested it with both yarn-client and yarn-cluster mode and things look fine.  I originally tried making `getApplicationReport` private[yarn], but discovered that it's needed by `YarnClientSchedulerBackend`.

Posting a new patch with the suggested changes.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: no, I meant move this up to the previous line if it doesn't exceed 100 chars. If it does, don't worry about it.

**Comment by andrewor14** [comment]
retest this please

**Comment by sryza** [comment]
The failures are legit.  Working on an update.

**Comment by sryza** [comment]
It looks like `copyFileToRemote` was a class method so that it could be mocked for a test (Mockito can't mock static methods).  There might be a more elegant way to write the test that doesn't require this, but I think that's probably out of the scope of this change, so I'm going to put it back for now.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Ah, cool.  Yeah, it doesn't fit on a single line.

**Comment by andrewor14** [comment]
Ok LGTM I'm merging this into master thanks

**Pull request closed without merging**